More direct message tests

anton-anufriev opened the pull request.

anton-anufriev committed:
More direct message tests

anton-anufriev committed:
More direct message tests

terzerm committed:
package moves, merged Message and DirectMessage, Factory now creates all DirectPayload objects

anton-anufriev committed:
More direct message tests

anton-anufriev committed:
More direct message tests

terzerm committed:
fix factory and clone command entries in InMemoryCommandLog

anton-anufriev committed:
Added Command.copyFrom() to copy command into a new CommandLogEntry
When appending CommandLogEntry to a CommandLog, CommandLog should set the next index to the CommandLogEntry and then append the entry to the end of the log.
Leader, on receiving commandMessage will construct a CommandLogEntry and append it the the log.

anton-anufriev committed:
Added Command.copyFrom() to copy command into a new CommandLogEntry
When appending CommandLogEntry to a CommandLog, CommandLog should set the next index to the CommandLogEntry and then append the entry to the end of the log.
Leader, on receiving commandMessage will construct a CommandLogEntry and append it the the log.

terzerm committed:
some more refactoring and cleanup around DirectFactory and Sender/Receiver

terzerm committed:
removed redundant truncateIncluding method from MessageLog

anton-anufriev committed:
Rebased with master.

anton-anufriev committed:
Merge remote-tracking branch 'origin/feature/anton_apr1' into feature/anton_apr1

# Conflicts:
#	src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java
#	src/main/java/org/tools4j/hoverraft/direct/AllocatingDirectFactory.java
#	src/main/java/org/tools4j/hoverraft/message/direct/DirectLogEntry.java

anton-anufriev committed:
Rebased with master.

anton-anufriev committed:
License comments added back.

anton-anufriev committed:
removed unused imports

terzerm reviewed (changes requested):
Please cherry-pick changes onto a new branch and create new pull request. Can't review like this, seeing all my changes.

anton-anufriev committed:
1. AppendResponse has serverId and matchLogEntryIndex.
2. TrackedFollowerState get initialised on new leadership.
3. LeaderState handles AppendResponse, updates nextIndex and matchIndex of Followers.
4. LeaderState implements this rule:
If there exists an N such that N > commitIndex, a majority
of matchIndex[i] ≥ N, and log[N].term == currentTerm:
set commitIndex = N (§5.3, §5.4).

anton-anufriev committed:
added readTo() method to read log entry into a logEntry instance.
some draft of sendAppendRequest method in Leaderstate.
